cli/src/lib/hashing/tokenHash.ts

```diff
@@ -0,0 +1,43 @@
+/**
+ * Generates a hashing object that can be used to create hashes of a
+ * token string. It is based on the rolling hash used in the
+ * Rabin-Karp string matching algorithm, but without the rolling
+ * property.
+ *
+ * The hash base is chosen to be different than the one used in
+ * RollingHash as to not create collisions with similar subsequent
+ * tokens (e.g. hash(['abc'] should not equal hash(['a', 'b', 'c'])).
+ */
+export class TokenHash {
+  /**
+   * The modulus used in the hash calculation.
+   *
+   * Since we want to be able to take the product of two hashes, the product of
+   * two hashes, the modulus multiplied by itself should not have more than the
+   * available amount of bits of precision.
+   *
+   * Javascript has 53-bit precision numbers (doubles) so. We pick the largest
+   * modulus m for which m * m has 53 bits.
+   */
+  readonly mod: number = 94906249;
+
+  /**
+   * The base (or radix) used in the hash calculation.
+   *
+   * We want that even small tokens use as much bits as possible. Because we
+   * assume the input token characters will mostly be less than 128, we pick
+   * our base such that multiplying with an input character use as much bits
+   * as possible.
+   *
+   * Hence this is the largest prime number such that 127 * base < mod.
+   */
+  readonly base: number = 747287;
+
+  public hashToken(token: string): number {
+    let hash = 0;
+    for (let i = 0; i < token.length; i++) {
+      hash = (this.base * hash + token.charCodeAt(i)) % this.mod;
```

Comment:
The hash of a single token is simply it's ascii value. This should be fixed.

Reply:
Yes, I was thinking this is not good. Any suggestions how to fix it?

Reply:
```javascript
      hash = (hash + token.charCodeAt(i)) * this.base % this.mod;
```

Reply:
This is may also be a problem for the rolling hash function.

Reply:
It is, but since the rolling hash will be used with the output hashes of the token hash, we should be using all the bits (with the fix you proposed).

Reply:
Another fix could be to add an initial (or final) step to multiply by a seed number which uses a lot of bits.